### PR TITLE
Virtual layer ui updates

### PR DESCRIPTION
**Pull request by rduivenvoorde**
See also: https://lists.osgeo.org/pipermail/qgis-developer/2021-August/063961.html

When testing the Virtual Layer functionality we had some UI issues. 

Some of them I try to address in this PR:

- the help button did not bring you to the documentation help
- the checkbox/input for the Unique Identifier column worked pretty quirky (also: WHAT is it for actually????):
  - you could uncheck it (with a value in the input) and then it still took it into account
  - the check for the fields validity was never reached
  - if you checked it you could still leave the field empty (but.. what happened?)
- when (re)defining the geometry, the default would result in a layer without CRS, you get asked for that now
- when you click the Add button, the same tests will be done as when you click the Test button (before you would just have an invalid layer)

One thing that I would(!) like to address, but could not find a way to:
- when you use the context menu 'update virtual layer' on an existing layer, the dialog reacts like you want to create a new one or a copy: asking you  a question on which you can only answer OK or Cancel 
It would be nice if you do the 'update', that you will just update the existing one (if you want to create a duplicate of the layer, use 'duplicate')
So: Question: IS there a way to distinguish between the 'creation of a new VL' and the 'update of a VL'? If so you could for example disable the layername part in the dialog or change the Add button caption to Ok or so...

### CODE REVIEW

**Review by asimrp — COMMENTED** [review]
This is the first time I'm reviewing a QGIS PR.  Please feel free to let me know if I'm not doing it right.

I didn't get a chance to test the patch but looks good upon reading.  Just a minor suggestion as posted above.

File: src/providers/virtual/qgsvirtuallayersourceselect.cpp [hunk 1]
```diff
@@ -229,13 +234,37 @@ void QgsVirtualLayerSourceSelect::testQuery()
         QMessageBox::warning( nullptr, tr( "Test Virtual Layer " ), tr( "The unique identifier field <b>%1</b> was not found in list of fields:<ul>%2</ul>" ).arg( mUIDField->text(), bulletedFieldNames.join( ' ' ) ) );
       }
       else
-        QMessageBox::information( nullptr, tr( "Test Virtual Layer" ), tr( "No error" ) );
+      {
+        if ( mGeometryRadio->isChecked() && mCRS->text().isEmpty() )
```
Comment (asimrp) [inline src/providers/virtual/qgsvirtuallayersourceselect.cpp:238]: This is a rather small function so please feel free to ignore the suggestion but I find it more readable when a function has only one entry and one exit point, as suggested below:
```suggestion
        bool result = false;
        ...
        result = (!(mGeometryRadio->isChecked() && mCRS->text().isEmpty()) ||
                  (QMessageBox::Yes == QMessageBox::question(...)))
        ...
        return result;
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
> when you use the context menu 'update virtual layer' on an existing layer, the dialog reacts like you want to create a new one or a copy: asking you a question on which you can only answer OK or Cancel
It would be nice if you do the 'update', that you will just update the existing one (if you want to create a duplicate of the layer, use 'duplicate')

Hmm... that's unfortunately not so straightforward. You'd basically need to move QgsVirtualLayerSourceSelect out to the GUI library instead of the virtual provider, so that you could then start refining that widget and refine it for editing existing layers. 

It would be a VERY welcome improvement though, if you do want to try!

**Comment by nyalldawson** [comment]
@rduivenvoorde let me know if you want to address @asimrp's comment or whether I should merge this now

File: src/providers/virtual/qgsvirtuallayersourceselect.cpp (hunk 1 above)
Comment (rduivenvoorde) [reply src/providers/virtual/qgsvirtuallayersourceselect.cpp:238]: @asimrp thanks for you review. I'm not a very experienced c++ coder, so I came up with this:

```
        result = ( !( mGeometryRadio->isChecked() && mCRS->text().isEmpty() ) ||
                   ( QMessageBox::Yes == QMessageBox::question( nullptr, tr( "Test Virtual Layer " ), tr( "No CRS defined, are you sure you want to create a layer without a crs?" ), QMessageBox::Yes | QMessageBox::No ) ) );
```
But although concise, I personally(!) cannot read this enough to see if the outcome is the same as the (indeed maybe too long if statements...
Not sure about real c++ coders, but I think I prefer to leave it like this (but others feel free to create cleaner code).

**Comment by rduivenvoorde** [comment]
> @rduivenvoorde let me know if you want to address @asimrp's comment or whether I should merge this now

See my comment. Others feel free to make this better, but I always get lost in statements with NOT's and OR's in it :-)


> Hmm... that's unfortunately not so straightforward. You'd basically need to move QgsVirtualLayerSourceSelect out to the GUI library instead of the virtual provider, so that you could then start refining that widget and refine it for editing existing layers.
> 
> It would be a VERY welcome improvement though, if you do want to try!

I want to try but then give me some time. If you want to clean up the PR list, then merge, else it is ok to let this open for now and I'll try in upcoming week.

**Comment by nyalldawson** [comment]
> If you want to clean up the PR list, then merge

Why yes, I very much do :tophat:

**Comment by rduivenvoorde** [comment]
> > If you want to clean up the PR list, then merge
> 
> Why yes, I very much do tophat

Ok, no problem :-)

Probably best anyway: I gave it a try to move the ui related files to other (gui) relevant places, but then missed knowledge to fix/update all the CMakeLists.txt files. Tried, but fell from one hole into another as I was not exactly sure what I was doing...
Maybe next hackfest ;-)

**Pull request merged**